test/rollover.cpp

```diff
@@ -0,0 +1,281 @@
+#include <stdx/rollover.hpp>
+
+#include <catch2/catch_template_test_macros.hpp>
+#include <catch2/catch_test_macros.hpp>
+
+#include <algorithm>
+#include <array>
+#include <cstdint>
+#include <iterator>
+#include <limits>
+#include <numeric>
+#include <type_traits>
+
+TEMPLATE_TEST_CASE("default construction", "[rollover]", std::uint8_t,
+                   std::uint16_t, std::uint32_t, std::uint64_t) {
+    using rollover_t = stdx::rollover_t<TestType>;
+    constexpr rollover_t x;
+    static_assert(x == rollover_t{TestType{}});
+    CHECK(x == rollover_t{TestType{}});
+}
+
+TEMPLATE_TEST_CASE("value construction", "[rollover]", std::uint8_t,
+                   std::uint16_t, std::uint32_t, std::uint64_t) {
+    using rollover_t = stdx::rollover_t<TestType>;
+    constexpr rollover_t x{};
+    static_assert(x == rollover_t{TestType{}});
+    CHECK(x == rollover_t{TestType{}});
+}
+
+TEMPLATE_TEST_CASE("access to underlying value", "[rollover]", std::uint8_t,
+                   std::uint16_t, std::uint32_t, std::uint64_t) {
+    using rollover_t = stdx::rollover_t<TestType>;
+    constexpr rollover_t x;
+    static_assert(x.as_underlying() == TestType{});
+    CHECK(x.as_underlying() == TestType{});
+}
+
+TEMPLATE_TEST_CASE("cast to underlying type", "[rollover]", std::uint8_t,
+                   std::uint16_t, std::uint32_t, std::uint64_t) {
+    using rollover_t = stdx::rollover_t<TestType>;
+    constexpr rollover_t x;
+    static_assert(static_cast<typename rollover_t::underlying_t>(x) ==
+                  TestType{});
+    CHECK(static_cast<typename rollover_t::underlying_t>(x) == TestType{});
+}
+
+TEMPLATE_TEST_CASE("construction from convertible integral value", "[rollover]",
+                   std::uint8_t, std::uint16_t, std::uint32_t, std::uint64_t) {
+    constexpr stdx::rollover_t<std::uint32_t> x{std::uint8_t{16}};
+    static_assert(x.as_underlying() == TestType{16});
+    CHECK(x.as_underlying() == TestType{16});
+}
+
+TEMPLATE_TEST_CASE("construction from convertible rollover_t", "[rollover]",
+                   std::uint8_t, std::uint16_t, std::uint32_t, std::uint64_t) {
+    constexpr stdx::rollover_t<std::uint32_t> x{
+        stdx::rollover_t<std::uint8_t>{16}};
+    static_assert(x.as_underlying() == TestType{16});
+    CHECK(x.as_underlying() == TestType{16});
+}
+
+TEST_CASE("conversion with deduction guide", "[rollover]") {
+    constexpr auto x = stdx::rollover_t{16};
+    static_assert(
+        std::is_same_v<decltype(x), stdx::rollover_t<unsigned int> const>);
+}
```

Comment:
suggestion: i don't think this should support automatic conversion from signed to unsigned. it should require an unsigned integral literal

Reply:
OK, fair.